jobs/anac-certified-attributes-importer/src/service/processor.ts

```diff
@@ -0,0 +1,218 @@
+import { ANAC_ABILITATO_CODE, ANAC_INCARICATO_CODE, ANAC_IN_CONVALIDA_CODE } from '../config/index.js'
+import { parse } from 'csv/sync'
+import {
+  CsvRow,
+  NonPaRow,
+  PaRow,
+  PersistentTenant,
+  InteropContext,
+  AnacAttributes,
+  BatchParseResult,
+  AttributeIdentifiers,
+} from '../model/index.js'
+import { ReadModelQueries, SftpClient, TenantProcessService } from '../service/index.js'
+import { RefreshableInteropToken, zipBy, logError, logWarn } from '@interop-be-reports/commons'
+import crypto from 'crypto'
+
+export async function importAttributes(
+  sftpClient: SftpClient,
+  readModel: ReadModelQueries,
+  tenantProcess: TenantProcessService,
+  refreshableToken: RefreshableInteropToken,
+  recordsBatchSize: number,
+  anacTenantId: string
+): Promise<void> {
+  const jobCorrelationId = crypto.randomUUID()
+  const batchSize = recordsBatchSize
+
+  const fileContent = await sftpClient.downloadCSV()
+
+  const attributes: AnacAttributes = await getAttributesIdentifiers(readModel, anacTenantId)
+
+  const preparedProcessTenants = processTenants(tenantProcess, refreshableToken, attributes, jobCorrelationId)
+
+  let scanComplete = false
+  let fromLine = 1
+
+  do {
+    const batchResult: BatchParseResult = getBatch(fileContent, fromLine, batchSize, jobCorrelationId)
+
+    // TODO Not sure this is the best way to satisfy type checks
+    const paOrgs: PaRow[] = batchResult.records
+      .map((org: CsvRow) => {
+        if ('codice_ipa' in org && org.codice_ipa) return org as PaRow
+        else return null
+      })
+      .filter((r): r is PaRow => r !== null)
+
+    const nonPaOrgs: NonPaRow[] = batchResult.records
+      .map((org: CsvRow) => {
+        if ('codice_ipa' in org) return null
+        else return org
+      })
+      .filter((r): r is NonPaRow => r !== null)
+    //
+
+    await preparedProcessTenants(
+      paOrgs,
+      (org) => org.codice_ipa,
+      (codes) => readModel.getPATenants(codes)
+    )
+    await preparedProcessTenants(
+      nonPaOrgs,
+      (org) => org.cf_gestore,
+      (codes) => readModel.getNonPATenants(codes)
+    )
+
+    fromLine = fromLine + batchSize
+    scanComplete = batchResult.processedRecordsCount === 0
+  } while (!scanComplete)
```

Comment:
as far as I got nodeJs has not the tail recursion optimization, so I avoided a recursive function 🥲